admiral/pkg/clusters/envoyfilter.go

```diff
@@ -0,0 +1,157 @@
+	"github.com/istio-ecosystem/admiral/admiral/pkg/controller/common"
+	log "github.com/sirupsen/logrus"
+	"istio.io/api/networking/v1alpha3"
+	networking "istio.io/client-go/pkg/apis/networking/v1alpha3"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"strings"
+)
+
+var (
+	getSha1 = common.GetSha1
+)
+func createOrUpdateEnvoyFilter( rc *RemoteController, routingPolicy *v1.RoutingPolicy, eventType admiral.EventType, workloadIdentityKey string, admiralCache *AdmiralCache) (*networking.EnvoyFilter, error) {
+
+	workloadSelectors := make(map[string]string)
+	workloadSelectors[common.GetWorkloadIdentifier()] = workloadIdentityKey
+	workloadSelectors[common.GetEnvKey()] = common.GetRoutingPolicyEnv(routingPolicy)
+
+	envoyfilterSpec := constructEnvoyFilterStruct(routingPolicy, workloadSelectors)
+
+	selectorLabelsSha, err := getSha1(workloadIdentityKey+common.GetRoutingPolicyEnv(routingPolicy))
+	if err != nil {
+		log.Error("error ocurred while computing workload labels sha1")
+		return nil, err
+	}
+	envoyFilterName := fmt.Sprintf("%s-dynamicrouting-%s-%s", strings.ToLower(routingPolicy.Spec.Plugin), selectorLabelsSha, "1.10")
+	envoyfilter := &networking.EnvoyFilter{
+		TypeMeta: metaV1.TypeMeta{
+			Kind:       "EnvoyFilter",
+			APIVersion: "networking.istio.io/v1alpha3",
+		},
+		ObjectMeta: metaV1.ObjectMeta{
+			Name:      envoyFilterName,
+			Namespace: common.NamespaceIstioSystem,
+		},
+		Spec: *envoyfilterSpec,
+	}
+
+	admiralCache.RoutingPolicyFilterCache.Put(workloadIdentityKey+common.GetRoutingPolicyEnv(routingPolicy), rc.ClusterID, envoyFilterName)
+	var filter *networking.EnvoyFilter
+	//get the envoyfilter if it exists. If it exists, update it. Otherwise create it.
+	if eventType == admiral.Add || eventType == admiral.Update {
+		// We query the API server instead of getting it from cache because there could be potential condition where the filter exists in the cache but not on the cluster.
+		filter, err = rc.RoutingPolicyController.IstioClient.NetworkingV1alpha3().EnvoyFilters(common.NamespaceIstioSystem).Get(envoyFilterName, metaV1.GetOptions{})
+		if err != nil {
+			log.Infof("msg=%s filtername=%s clustername=%s", "creating the envoy filter", envoyFilterName, rc.ClusterID)
+			filter, err = rc.RoutingPolicyController.IstioClient.NetworkingV1alpha3().EnvoyFilters(common.NamespaceIstioSystem).Create(envoyfilter)
+			if err != nil {
+				log.Infof("error creating filter: %v", err)
+			}
+		} else {
+			log.Infof("msg=%s filtername=%s clustername=%s", "updating existing envoy filter", envoyFilterName, rc.ClusterID)
+			envoyfilter.ResourceVersion = filter.ResourceVersion
+			filter, err = rc.RoutingPolicyController.IstioClient.NetworkingV1alpha3().EnvoyFilters(common.NamespaceIstioSystem).Update(envoyfilter)
+		}
+	}
+
+
+	return filter, err
+}
+
+func constructEnvoyFilterStruct(routingPolicy *v1.RoutingPolicy, workloadSelectorLabels map[string]string) *v1alpha3.EnvoyFilter {
```

Comment:
Can we make this more generic, so that we can easily use this for a different version of envoy.